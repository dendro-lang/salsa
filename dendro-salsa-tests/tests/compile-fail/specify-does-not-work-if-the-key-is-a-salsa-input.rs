//! Test that `specify` does not work if the key is a `salsa::input`
//! compilation fails
#![allow(warnings)]

#[salsa::jar(db = Db)]
struct Jar(MyInput, MyTracked, tracked_fn);

trait Db: salsa::DbWithJar<Jar> {}

#[salsa::input(jar = Jar)]
struct MyInput {
    field: u32,
}

#[salsa::tracked(jar = Jar)]
struct MyTracked {
    field: u32,
}

#[salsa::tracked(jar = Jar, specify)]
fn tracked_fn(db: &dyn Db, input: MyInput) -> MyTracked {
    MyTracked::new(db, input.field(db) * 2)
}

#[salsa::db(Jar)]
#[derive(Default)]
struct Database {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for Database {}

impl Db for Database {}

fn main() {}
