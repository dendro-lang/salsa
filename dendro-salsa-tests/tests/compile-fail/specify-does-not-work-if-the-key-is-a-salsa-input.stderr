error[E0277]: the trait bound `MyInput: TrackedStructInDb<dyn Db>` is not satisfied
  --> tests/compile-fail/specify-does-not-work-if-the-key-is-a-salsa-input.rs:20:1
   |
20 | #[salsa::tracked(jar = Jar, specify)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `TrackedStructInDb<dyn Db>` is not implemented for `MyInput`
   |
   = help: the trait `TrackedStructInDb<DB>` is implemented for `MyTracked`
note: required by a bound in `function::specify::<impl FunctionIngredient<C>>::specify_and_record`
  --> $WORKSPACE/components/dendro-salsa/src/function/specify.rs
   |
   |     pub fn specify_and_record<'db>(&self, db: &'db DynDb<'db, C>, key: C::Key, value: C::Value)
   |            ------------------ required by a bound in this associated function
   |     where
   |         C::Key: TrackedStructInDb<DynDb<'db, C>>,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `function::specify::<impl FunctionIngredient<C>>::specify_and_record`
   = note: this error originates in the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)
